game/src/edit/new_traffic_signals.rs

```diff
@@ -0,0 +1,766 @@
+use crate::app::{App, ShowEverything};
+use crate::common::CommonState;
+use crate::edit::traffic_signals::{draw_selected_group, make_top_panel, PreviewTrafficSignal};
+use crate::game::{ChooseSomething, DrawBaselayer, State, Transition};
+use crate::options::TrafficSignalStyle;
+use crate::render::{draw_signal_phase, DrawOptions, DrawTurnGroup};
+use crate::sandbox::{spawn_agents_around, GameplayMode};
+use abstutil::Timer;
+use ezgui::{
+    hotkey, Btn, Checkbox, Choice, Color, Composite, Drawable, EventCtx, GeomBatch, GfxCtx,
+    HorizontalAlignment, Key, Line, Outcome, Spinner, Text, TextExt, VerticalAlignment, Widget,
+};
+use geom::{Bounds, Distance, Duration, Polygon};
+use map_model::{
+    ControlTrafficSignal, IntersectionID, Phase, PhaseType, TurnGroupID, TurnPriority,
+};
+use std::collections::BTreeSet;
+
+// Welcome to one of the most overwhelmingly complicated parts of the UI...
+
+pub struct NewTrafficSignalEditor {
+    side_panel: Composite,
+    top_panel: Composite,
+
+    gameplay: GameplayMode,
+    members: BTreeSet<IntersectionID>,
+    current_phase: usize,
+
+    groups: Vec<DrawTurnGroup>,
+    // And the next priority to toggle to
+    group_selected: Option<(TurnGroupID, Option<TurnPriority>)>,
+    draw_current: Drawable,
+
+    // The first is the original
+    command_stack: Vec<BundleEdits>,
+    redo_stack: Vec<BundleEdits>,
+
+    fade_irrelevant: Drawable,
+}
+
+// For every member intersection, the full state of that signal
+#[derive(Clone)]
+struct BundleEdits {
+    signals: Vec<ControlTrafficSignal>,
+}
+
+impl NewTrafficSignalEditor {
```

Comment:
I'm not sure how much of this was copied from the old signal editor, so unsure which parts to scrutinize more. LMK if there's something you'd like me to especially look at.